### PR TITLE
Respect Solidity version when typechecking builtin constructs

### PR DESCRIPTION
**Pull request by blitz-1306**
## Preface
Current logic of typechecking is a bit permissive when developer uses builtins in predicates. For an instance currently developer able to use `address(0x0).code` ([introduced in Solidity 0.8.0](https://github.com/ethereum/solidity/releases/tag/v0.8.0)) for Solidity 0.6.0. Scribble can be tuned to respect Solidity version during typechecking.

## Changes
- [x] Tweaked `TypeEnv` to be aware of target compiler version.
- [x] Tweaked `BuiltinStructType`, `BuiltinSymbols` and `BuiltinAddressMembers` to allow specifying base required compiler version for builtin constructs.
- [x] Tweaked typecheck logic to respect required Solidity version of builtin constructs.
- [x] Tweaked tests to indicate the difference.

## Notes
This PR introduces logic for such checks. Newer builtins would be introduced in other further PRs.

Regards.

### CODE REVIEW

File: src/spec-lang/tc/typecheck.ts [hunk 1]
```diff
@@ -1355,17 +1390,18 @@ export function tcMemberAccess(expr: SMemberAccess, ctx: STypingCtx, typeEnv: Ty
             }
         }
 
-        const type = BuiltinAddressMembers.get(expr.member);
+        // const type = BuiltinAddressMembers.get(expr.member);
```
Comment (cd1m0) [inline src/spec-lang/tc/typecheck.ts:1393]: Commented code got left here :)

File: src/spec-lang/tc/typecheck.ts [hunk 2]
```diff
@@ -1297,17 +1322,27 @@ export function tcMemberAccess(expr: SMemberAccess, ctx: STypingCtx, typeEnv: Ty
...
+        if (result === undefined) {
             throw new SNoField(
                 `Builtin struct "${expr.base.pp()}" does not have a member "${expr.member}"`,
                 expr,
                 expr.member
             );
         }
 
-        return type;
+        if (result instanceof Array) {
```
Comment (cd1m0) [inline src/spec-lang/tc/typecheck.ts:1335]: There is some code duplication between this and `tcIdBuiltinSymbol`. Can we abstract in a helper function?

**Review by cd1m0 — CHANGES_REQUESTED** [review]
Good work! Two small nits.

File: src/spec-lang/tc/typecheck.ts (hunk 1 above)
Comment (blitz-1306) [reply src/spec-lang/tc/typecheck.ts]: Addressed in https://github.com/ConsenSys/scribble/pull/45/commits/3fda5277eb180bda7d656fc0d036f878c21290fe

File: src/spec-lang/tc/typecheck.ts (hunk 2 above)
Comment (blitz-1306) [reply src/spec-lang/tc/typecheck.ts]: Addressed in https://github.com/ConsenSys/scribble/pull/45/commits/3fda5277eb180bda7d656fc0d036f878c21290fe

**Review by cd1m0 — APPROVED** [review]
LGTM

**Pull request merged**